src/ini.cr

```diff
@@ -1,22 +1,47 @@
 class INI
+  # Exception thrown on an INI parse error.
+  class ParseException < Exception
+    getter line_number : Int32
+    getter column_number : Int32
+
+    def initialize(message, @line_number, @column_number)
+      super "#{message} at #{@line_number}:#{@column_number}"
+    end
+
+    def location
+      {line_number, column_number}
+    end
+  end
+
   # Parses INI-style configuration from the given string.
+  # Raises a `ParseException` on any errors.
   #
   # ```
   # INI.parse("[foo]\na = 1") # => {"foo" => {"a" => "1"}}
   # ```
   def self.parse(str) : Hash(String, Hash(String, String))
     ini = {} of String => Hash(String, String)
-
+    lines = str.lines.map(&.lstrip)
     section = ""
-    str.each_line do |line|
-      if line =~ /\s*(.*[^\s])\s*=\s*(.*[^\s])/
-        ini[section] ||= {} of String => String if section == ""
-        ini[section][$1] = $2
-      elsif line =~ /\[(.*)\]/
-        section = $1
+
+    lines.each_with_index(1) do |line, lineno|
+      next if line.empty?
+      next if line[0] == '#' || line[0] == ';'
+
+      if line[0] == '['
+        end_idx = line.index(']')
+        raise ParseException.new("unterminated section", lineno, line.size - 1) unless end_idx
```

Comment:
This should probably also raise if `end_idx` is not the last character in the line. `[foo] bar` should be invalid.